src/common.rkt

```diff
@@ -104,6 +101,81 @@
   (check-true (set-disjoint? '() '()))
   (check-false (set-disjoint? '(a b c) '(a))))
 
+(define (list-suffix? l r)
+  (list-prefix? (reverse l) (reverse r)))
+
+(module+ test
+  (check-true (list-suffix? empty empty))
+  (check-true (list-suffix? '(1 2) '(0 1 2)))
+  (check-false (list-suffix? '(1 2 3) '(0 1 2)))
+  (check-false (list-suffix? '(1 2 3) '())))
+
+(define (subsequence? v l)
+  (or
+   (empty? v)
+   (let ([v* (member (first v) l)])
+     (and v* (subsequence? (rest v) v*)))))
+
+(module+ test
+  (define l (range 10))
+  (check-true (subsequence? empty empty))
+  (check-true (subsequence? empty l))
+  (check-true (subsequence? '(1) l))
+  (check-true (subsequence? '(1 2) l))
+  (check-true (subsequence? '(1 3 5 7 9) l))
+  (check-true (subsequence? '(1 2 5 8) l))
+  (check-false (subsequence? '(x y) l))
+  (check-false (subsequence? '(1 2 10) l)))
+
+(define (list-ref* l p)
+  (let loop ([l l] [p p] [i 0])
+    (if (or (empty? p) (empty? l))
+        empty
+        (let* ([j (first p)]
+               [k (- j i)]
+               [l* (drop l k)])
+          (cons (first l*) (loop (rest l*) (rest p) (+ i k 1)))))))
+
+(module+ test
+  (define m '(a b c d e f g))
+  (check-equal? (list-ref* empty empty) empty)
+  (check-equal? (list-ref* m empty) empty)
+  (check-equal? (list-ref* m '(1)) '(b))
+  (check-equal? (list-ref* m '(0 2 4 6)) '(a c e g))
+  (check-equal? (list-ref* m '(0 2 3 5 6)) '(a c d f g)))
+
+(define (list-set* l p v)
+  (let loop ([l l] [p p] [v v] [i 0])
+    (cond
+      [(empty? l)
+       empty]
+      [(and (not (empty? p)) (equal? (first p) i))
+       (cons (first v) (loop (rest l) (rest p) (rest v) (add1 i)))]
+      [else
+       (cons (first l) (loop (rest l) p v (add1 i)))])))
+
+(module+ test
+  (define n '(a b c d e f g))
+  (check-equal? (list-set* empty empty empty) empty)
+  (check-equal? (list-set* n empty empty) n)
+  (check-equal? (list-set* n '(0) '(x)) '(x b c d e f g))
+  (check-equal? (list-set* n '(1 2 5) '(x y z)) '(a x y d e z g)))
+
+;; Union-find
+
+(define (disjoint-set s)
+  (list->vector (range s)))
+
+(define (disjoint-set-find! d x)
+  (define p (vector-ref d x))
+  (if (= p x)
+      x
+      (let ([g (vector-ref d p)])
+        (vector-set! d x g)
```

Comment:
I don't think this is actually how path-compression works. You should do a `find!` to get the ultimate representative, then save that.

Reply:
Yeah I tried to implement the path splitting that the Wikipedia page for Union-Find mentions, but I wasn't sure I had totally wrapped my head around it. I think it should be fixed now.